src/tribler/core/components/database/db/layers/user_activity_layer.py

```diff
@@ -0,0 +1,128 @@
+from __future__ import annotations
+
+import random
+import typing
+from dataclasses import dataclass
+
+from pony import orm
+from pony.orm import db_session
+
+from tribler.core.components.user_activity.types import InfoHash
+from tribler.core.utilities.pony_utils import TrackedDatabase
+
+if typing.TYPE_CHECKING:
+    @dataclass
+    class InfohashPreference:
+        infohash: bytes
+        preference: float
+        parent_query: Query
+
+    @dataclass
+    class Query:
+        query: str
+        infohashes: typing.Set[InfohashPreference]
+
+
+class UserActivityLayer:
+
+    def __init__(self, database: TrackedDatabase, update_weight: float = 0.8, e: float = 0.01) -> None:
+        """
+        Create a new User Activity scheme for a particular database.
+
+        :param database: The database to bind to.
+        :param update_weight: The weight of new updates.
+        :param e: A small value to decide near-zero preference.
+        """
+        self.database = database
+
+        self.e = e
+        self.update_weight_new = update_weight
+        self.update_weight_old = 1 - self.update_weight_new
+
+        class Query(database.Entity):
+            query = orm.PrimaryKey(str)
+            infohashes = orm.Set("InfohashPreference")
+
+        class InfohashPreference(database.Entity):
+            infohash = orm.Required(bytes)
+            preference = orm.Required(float)
+            parent_query = orm.Required(Query)
+            orm.PrimaryKey(infohash, parent_query)
```

Comment:
The database does not appear to be normalized. As @kozlovsky is coming back from vacation at the same time as you, I summon him to review the database structure.

Reply:
After some additional thought, I agree with @drew2a that it may be better to link the `InfohashPreference` entity with the `Resource` entity in the same way as the `Tracker` entity of the `HealthDataAccessLayer` is linked with the `Resource` entity via the `Tracker.torrents` / `Resource.trackers` relationships.

It has the following benefits:
1. The primary key of `InfohashPreference` becomes shorter, as it now will use resource id instead of infohash bytes.
2. With the current approach of `TriblerDatabase`, the torrent metadata is just a kind of `Resource`, and having `Resource` directly linked with `InfohashPreference` can simplify some future queries.

The drawback is when we want to search `InfohashPreference` knowing the specific info hash, we first need to find the `Resource` and then use it instead of the info hash value.

With this change, the code will be like:
```python
class UserActivityLayer:

    def __init__(self, knowledge_layer: KnowledgeDataAccessLayer) -> None:
        self.database = knowledge_layer.instance
        self.Resource = knowledge_layer.Resource

        class Query(self.database.Entity):
            query = orm.PrimaryKey(str)
            searched_counter = orm.Required(int, default=1)
            infohashes = orm.Set("InfohashPreference")

        class InfohashPreference(self.database.Entity):
            torrent = orm.Required(self.Resource)
            query = orm.Required(Query)
            chosen_counter = orm.Required(int, default=0)
            ignored_counter = orm.Required(int, default=0)
            orm.PrimaryKey(torrent, query)

        self.Query = Query
        self.UserPreference = InfohashPreference
```

(In this example, I renamed `parent_query` to `query`, as `parent_` prefix looks unnecessary)

And, in the `Resource` entity of the `KnowledgeDataAccessLayer` we will have:
```python
infohash_preferences = orm.Set(lambda: db.InfohashPreference, reverse="torrent")
```

Reply:
I don't see how you could implement decay of previous search and results for the same (infohash, query) with this database format. Is it still possible? Because that is a requirement.

Reply:
Indeed, it is not enough to implement the proper decay; I missed that requirement. But it leads me to some additional thoughts.

The current scheme implemented in this PR is single-user. I don't think decay is important when the statistics are accumulated only for a single user. But if we aggregate anonymized query-result-preference statistics about thousands of users, the decay indeed makes sense.

But then we have a new question on spreading and accumulating these statistics. It probably should be signed by a second key when gossiping to prevent spam. However, we cannot sign the dynamically decaying `preference` value of the `float` type. We can sign some discrete information that at the moment T, an anonymous user U performed the query Q and clicked on the infohash H.

So, if the goal is to aggregate decaying anonymous user clicks-at-query-results statistics across multiple users, the discrete signed piece of information should probably be (T, U, Q, H). Then, the decay can be implemented by taking the time stamps into account - the weight of the user's "vote" can be inversely proportional to the vote's age.

In that case, the entity attributes might be something like:

```python
class InfohashPreference(self.database.Entity):
    user = orm.Required(User)  # some new entity
    query = orm.Required(Query)
    torrent = orm.Required(self.Resource)
    last_clicked_at = orm.Required(datetime)
    signature = orm.Required(bytes)
    orm.PrimaryKey(user, query, torrent)
```

What do you think?

Reply:
That is close to what I had in mind for the long term, in a different PR. I would prefer we discuss the grand design in the linked issue, not on this PR.

Just to touch on it, in short, the plan for now is to use emergent behavior, as follows:
1. Torrents that are "preferred" have their health checked more frequently locally (this PR).
2. Torrents that have their health checked recently (locally) are more frequently gossiped in the popularity community (already exists).
3. Torrents that are gossiped more by others have a higher chance of appearing in search results remotely (already exists).
4. _Effect:_ search results that are gossiped to more users are more likely to be downloaded. For actually popular content, that is downloaded, this forms a feedback loop: back to step 2.

In summary, this PR creates a soft bias and, therefore, an emergent effect that boosts the popularity of content that is searched for and downloaded.

Establishing shadow identities and more aggressively gossiping - while preventing spam - is something I'll leave for a follow-up PR. Ideally, we don't need to gossip preference directly and we can somehow merge gossiped ML models. However, this should only be implemented after careful experimentation. For now, this PR gives each user a local ranking to start the ML experimentation.

Reply:
Thanks for the explanation; now I understand your approach better. Initially, I was misguided by the picture in #7632 with a "store signed" label, as my understanding was that it is only possible to sign discrete facts, not float values. If gossiping about individual provable facts is not the intention, then storing calculated preferences is probably fine.

Still, you can consider using `torrent = orm.Required(self.Resource)` instead of `infohash = orm.Required(bytes)` in the `InfohashPreference` entity to reduce the data size.

Reply:
Sure, thank you for the suggestion. Once the initial prototype has been merged, we can look at refactoring and optimizations and I will definitely keep your suggestion in mind. That said, once this has been merged, this PR is also no longer my sole responsibility and others can also contribute their excellent suggestions to the communal code. We can grow the code over time.

I do realize, now, that I left `enabled = True` as the default setting. I'll make sure to keep this disabled by default so we still have the freedom to change things like the database format in future PRs.